# Đồ án cuối kỳ



|    Họ và Tên    |    MSSV    |
| ---------|----------|
| Tạ Tiến Thành Đạt| 1712333    |
| Nguyễn Nhật Hảo| 19120501   |

Link Trello: https://trello.com/b/udLhkMpv/l%E1%BB%8Bch-tr%C3%ACnh

## Import các thư viện cần thiết

In [1]:
import numpy as np
import pandas as pd
import matplotlib as plt

## Thu thập dữ liệu

Trong đồ án này chúng em sử dụng [bộ dữ liệu](https://www.kaggle.com/altadata/covid19?fbclid=IwAR3bOuHzXvun8UHQd-6xKWAxpz9m_hICvuBF0zXkbWNh4TlorFre7_GyV40) về COVID-19 được public trên Kaggle. Bộ dữ liệu này ghi lại những thông tin số liệu thống kê về đại dịch COVID-19 hằng ngày trên toàn thế giới do Trung tâm Khoa học Hệ thống và Kỹ thuật (CSSE) của Đại học Johns Hopkins (JHU) cung cấp. Sở dĩ nhóm chúng em chọn bộ dữ liệu này cho đồ án bởi như tất cả mọi người đều đã thấy, sự ảnh hưởng mà đại dịch COVID-19 đã gây ra cho thế giới trong hơn 2 năm qua thật sự to lớn. Và chính bản thân những người sinh viên như chúng em cũng đã phải chịu ảnh hưởng nhất định đến việc học việc làm cũng như đời sống. Đó cũng chính là nguồn động lực thúc đẩy chúng em muốn tìm hiểu và phân tích nhiều hơn về đại dịch trong đồ án này.

## Khám phá dữ liệu và tiền xử lý

**Đọc dữ liệu từ file csv vào dataframe**

In [2]:
df = pd.read_csv('Data/altadatacovid_19.csv')
df

reported_date                  country_region  population  confirmed  \
0        2020-02-24                     Afghanistan  38928341.0          1   
1        2020-02-25                     Afghanistan  38928341.0          1   
2        2020-02-26                     Afghanistan  38928341.0          1   
3        2020-02-27                     Afghanistan  38928341.0          1   
4        2020-02-28                     Afghanistan  38928341.0          1   
...             ...                             ...         ...        ...   
32606    2020-03-12  occupied Palestinian territory         NaN          0   
32607    2020-03-14  occupied Palestinian territory         NaN          0   
32608    2020-03-15  occupied Palestinian territory         NaN          0   
32609    2020-03-16  occupied Palestinian territory         NaN          0   
32610    2020-03-17  occupied Palestinian territory         NaN          0   

       active  deaths  recovered  mortality_rate  incident_rate  
0           1       0          0             0.0          0.003  
1           1       0          0             0.0          0.003  
2           1       0          0             0.0          0.003  
3           1       0          0             0.0          0.003  
4           1       0          0             0.0          0.003  
...       ...     ...        ...             ...            ...  
32606       0       0          0             NaN            NaN  
32607       0       0          0             NaN            NaN  
32608       0       0          0             NaN            NaN  
32609       0       0          0             NaN            NaN  
32610       0       0          0             NaN            NaN  

[32611 rows x 9 columns]

**Dữ liệu có bao nhiêu dòng và bao nhiêu cột?**

In [4]:
num_rows, num_cols = df.shape
print('số dòng: ', num_rows)
print('số cột: ', num_cols)

số dòng:  32611
số cột:  9


**Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa
khác nhau không?**

- Theo mô tả và cũng như những gì quan sát được thì mỗi dòng dữ liệu sẽ là một ghi nhận số liệu về thống kê liên quan tới covid mỗi ngày ở mỗi quốc gia. Chưa thấy có điều bất thường nào quá rõ ràng nên ta cần khám phá và quan sát thêm để hiểu thêm về dữ liệu.

**Dữ liệu có các dòng bị lặp không??**

In [9]:
if True in df.duplicated().unique():
    print('Dữ liệu có các dòng bị lặp')
else:
    print('Dữ liệu không có dòng lặp')

Dữ liệu có các dòng bị lặp


- *quan sát dòng bị lặp*

In [7]:
df[df.duplicated()]

reported_date country_region  population  confirmed  active  deaths  \
24084    2020-03-16          Congo   5518092.0          1       1       0   

       recovered  mortality_rate  incident_rate  
24084          0             0.0          0.018

- Ta sẽ xóa luôn dòng dữ liệu bị lặp này

In [100]:
df.drop_duplicates()

reported_date country_region  population  confirmed  active  deaths  \
0        2020-02-24    Afghanistan  38928341.0          1       1       0   
1        2020-02-25    Afghanistan  38928341.0          1       1       0   
2        2020-02-26    Afghanistan  38928341.0          1       1       0   
3        2020-02-27    Afghanistan  38928341.0          1       1       0   
4        2020-02-28    Afghanistan  38928341.0          1       1       0   
...             ...            ...         ...        ...     ...     ...   
32599    2020-08-19       Zimbabwe  14862927.0       5643    1051     150   
32600    2020-08-20       Zimbabwe  14862927.0       5745    1069     151   
32601    2020-08-21       Zimbabwe  14862927.0       5815    1076     152   
32602    2020-08-22       Zimbabwe  14862927.0       5893    1111     153   
32603    2020-08-23       Zimbabwe  14862927.0       5930     903     155   

       recovered  mortality_rate  incident_rate  
0              0           0.000          0.003  
1              0           0.000          0.003  
2              0           0.000          0.003  
3              0           0.000          0.003  
4              0           0.000          0.003  
...          ...             ...            ...  
32599       4442           2.658         37.967  
32600       4525           2.628         38.653  
32601       4587           2.614         39.124  
32602       4629           2.596         39.649  
32603       4872           2.614         39.898  

[31465 rows x 9 columns]

**Mỗi cột có ý nghĩa gì?**

Các cột của bộ dữ liệu được mô tả trên Kaggle như sau:

- reported_date: Covid-19 Report Date

- country_region: Country, region or sovereignty name

- population: Country populations as per United Nations Population Division

- confirmed: Confirmed cases include presumptive positive cases and probable cases

- active: Active cases = total confirmed - total recovered - total deaths

- deaths: Death cases counts

- recovered: Recovered cases counts

- mortality_rate: Number of recorded deaths * 100 / Number of confirmed cases

- incident_rate: Confirmed cases per 100,000 persons

**Mỗi cột hiện đang có kiểu dữ liệu gì? Có cột nào có kiểu
dữ liệu chưa phù hợp để có thể xử lý tiếp không?**

+ Ta sẽ quan sát kiểu dữ liệu của các cột

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32611 entries, 0 to 32610
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   reported_date   32611 non-null  object 
 1   country_region  32611 non-null  object 
 2   population      31466 non-null  float64
 3   confirmed       32611 non-null  int64  
 4   active          32611 non-null  int64  
 5   deaths          32611 non-null  int64  
 6   recovered       32611 non-null  int64  
 7   mortality_rate  32561 non-null  float64
 8   incident_rate   31466 non-null  float64
dtypes: float64(3), int64(4), object(2)
memory usage: 2.2+ MB


- Cột reported_date đang có kiểu dữ liệu là object nên ta sẽ chuyển về kiểu datime

In [14]:
df['reported_date'] = pd.to_datetime(df['reported_date'], format='%Y-%m-%d')

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32611 entries, 0 to 32610
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   reported_date   32611 non-null  datetime64[ns]
 1   country_region  32611 non-null  object        
 2   population      31466 non-null  float64       
 3   confirmed       32611 non-null  int64         
 4   active          32611 non-null  int64         
 5   deaths          32611 non-null  int64         
 6   recovered       32611 non-null  int64         
 7   mortality_rate  32561 non-null  float64       
 8   incident_rate   31466 non-null  float64       
dtypes: datetime64[ns](1), float64(3), int64(4), object(1)
memory usage: 2.2+ MB


* Ta thấy có 3 cột dữ liệu bị thiếu giá trị là population, mortality_rate và incident_rate. Trong đó số giá trị bị thiếu ở 2 cột population và incident_rate là ngang nhau. Có thể trong việc này có mỗi liên hệ với nhau. Bên cạnh đó, cột mortality chỉ bị thiếu khá ít dữ liệu. Nhưng ta sẽ tạm chưa bàn tới việc này và để chúng lại ở phần tiền xử lý giá trị thiếu.

**Với mỗi cột có kiểu dữ liệu dạng số (numerical), các giá
trị được phân bố như thế nào?**

- Theo mô tả về dữ liệu và quan sát về kiểu dữ liệu của các cột thì chỉ có cột country_region là thuộc kiểu dữ liệu categorical. Còn lại tất cả các cột khác ta sẽ đều xem xét chúng ở dạng số 

In [103]:
def missing_ratio(s):
    return s.isna().mean() * 100

In [17]:
numerical = ['reported_date', 'population', 'confirmed', 'active',
       'deaths', 'recovered', 'mortality_rate', 'incident_rate']

In [25]:
df[numerical].describe()

population     confirmed        active         deaths     recovered  \
count  3.146600e+04  3.261100e+04  3.261100e+04   32611.000000  3.261100e+04   
mean   4.497319e+07  4.205268e+04  1.775490e+04    1934.022140  2.236375e+04   
std    1.573180e+08  2.437990e+05  1.340613e+05    9898.133659  1.228619e+05   
min    8.090000e+02  0.000000e+00 -2.000000e+00       0.000000  0.000000e+00   
25%    2.722291e+06  9.900000e+01  3.100000e+01       1.000000  1.800000e+01   
50%    1.009927e+07  1.195000e+03  4.220000e+02      23.000000  4.400000e+02   
75%    3.286627e+07  1.068850e+04  3.695500e+03     239.000000  4.513500e+03   
max    1.439324e+09  5.701679e+06  3.527116e+06  176802.000000  2.947784e+06   

       mortality_rate  incident_rate  
count    32561.000000   31466.000000  
mean         3.219820     150.381521  
std          4.528614     344.922692  
min          0.000000       0.000000  
25%          0.476000       2.488500  
50%          1.917000      23.177500  
75%          4.211000     129.171750  
max        100.000000    4061.283000

- Tỉ lệ giá trị thiếu

In [104]:
missing_ratio(df[numerical])

reported_date     0.0
population        0.0
confirmed         0.0
active            0.0
deaths            0.0
recovered         0.0
mortality_rate    0.0
incident_rate     0.0
dtype: float64

**Với mỗi cột có kiểu dữ liệu dạng phân loại (categorical),
các giá trị được phân bố như thế nào?**

- Tỉ lệ giá trị thiếu

In [105]:
missing_ratio(df[categorical])

country_region    0.0
dtype: float64

In [31]:
categorical = ['country_region']
df[categorical].value_counts()

country_region          
Republic of Korea           215
United States of America    215
Thailand                    215
Japan                       215
Viet Nam                    214
                           ... 
Ivory Coast                   1
Taipei and environs           1
Cape Verde                    1
North Ireland                 1
China, Macao SAR              1
Length: 220, dtype: int64

- Như vậy có 220 quốc gia được ghi nhận số liệu thống kê trong bộ dữ liệu về covid này. Trong đó những quốc gia như Hàn Quốc, Mỹ,... được ghi nhận với nhiều dòng dữ liệu nhất. Và có một số quốc gia thì ngược lại chỉ có một dòng dữ liệu được ghi nhận. Như những gì ta đã được quan sát thì mỗi dòng dữ liệu cũng tương ứng với một ngày mà dữ liệu về covid ở mỗi quốc gia được ghi nhận.

***Xử lý giá trị thiếu***

- Giờ ta trở lại với 3 cột dữ liệu có giá trị thiếu đã khám phá ở phía trên. Ta sẽ tìm cách để giải quyết những chỗ bị thiếu này. Nhớ lại những gì đã khám phá và phần mô tả cho dữ liệu thì cột population là dân số của quốc gia, còn cột incident_rate là tỉ lệ số ca mắc được xác nhận trên 100,000 ca. Như vậy có thể việc số giá trị thiếu ở 2 cột này giống nhau là do không có số liệu về dân số để có thể tính được tỉ lệ số ca mắc theo dân số.

- Nếu bằng một cách nào đó tính toán dựa trên dữ liệu để điền vào những giá trị bị thiếu về dân số thì có vẻ sẽ không được khách quan lắm. Nên ở đây em chọn phương án bỏ đi những dòng dữ liệu có cột dân số bị thiếu.

In [72]:
df = df.dropna(subset=['population', 'incident_rate'])
df

reported_date country_region  population  confirmed  active  deaths  \
0        2020-02-24    Afghanistan  38928341.0          1       1       0   
1        2020-02-25    Afghanistan  38928341.0          1       1       0   
2        2020-02-26    Afghanistan  38928341.0          1       1       0   
3        2020-02-27    Afghanistan  38928341.0          1       1       0   
4        2020-02-28    Afghanistan  38928341.0          1       1       0   
...             ...            ...         ...        ...     ...     ...   
32599    2020-08-19       Zimbabwe  14862927.0       5643    1051     150   
32600    2020-08-20       Zimbabwe  14862927.0       5745    1069     151   
32601    2020-08-21       Zimbabwe  14862927.0       5815    1076     152   
32602    2020-08-22       Zimbabwe  14862927.0       5893    1111     153   
32603    2020-08-23       Zimbabwe  14862927.0       5930     903     155   

       recovered  mortality_rate  incident_rate  
0              0           0.000          0.003  
1              0           0.000          0.003  
2              0           0.000          0.003  
3              0           0.000          0.003  
4              0           0.000          0.003  
...          ...             ...            ...  
32599       4442           2.658         37.967  
32600       4525           2.628         38.653  
32601       4587           2.614         39.124  
32602       4629           2.596         39.649  
32603       4872           2.614         39.898  

[31466 rows x 9 columns]

- Không có gì bất thường.

- Ở cột dữ liệu còn lại bị thiếu là mortality_rate (tỉ lệ tử vong trên số ca mắc) ta có thể tính toán được bằng cách chia dữ liệu ở cột deaths (số ca tử vong) cho cột confirmed (số ca mắc được xác nhận)

*Tuy nhiên khi quan sát lại thì ta thấy ở những dòng còn lại mà bị thiếu giá trị ở cột mortaity thì những cột thông kê khác có liên quan đều bằng 0. Tới đây có 2 suy nghĩ khả dĩ là xóa luôn những dòng bị thiếu này hoặc điền giá trị 0 cho những cột thiếu. Và do cá nhân em cảm giác những dòng này hoàn toàn có thể sẽ cần dùng đến nên em chọn phương án điền vào giá trị 0 cho những ô bị thiếu này*

In [79]:
df[df['mortality_rate'].isna()]

reported_date        country_region   population  confirmed  active  \
1384     2020-01-23             Australia   25499881.0          0       0   
3975     2020-01-23                Brazil  212559409.0          0       0   
6370     2020-01-23              Colombia   50882884.0          0       0   
11864    2020-03-19             Greenland      56772.0          0       0   
11865    2020-03-20             Greenland      56772.0          0       0   
11866    2020-03-21             Greenland      56772.0          0       0   
12033    2020-03-18                  Guam     168783.0          0       0   
12034    2020-03-19                  Guam     168783.0          0       0   
12035    2020-03-20                  Guam     168783.0          0       0   
12036    2020-03-21                  Guam     168783.0          0       0   
13179    2020-01-22  China, Hong Kong SAR    7496988.0          0       0   
14148    2020-02-23                  Iraq   40222503.0          0       0   
18464    2020-01-23              Malaysia   32365998.0          0       0   
19511    2020-01-23                Mexico  128932753.0          0       0   
23342    2020-01-23           Philippines  109581085.0          0       0   
23899    2020-03-17           Puerto Rico    2860840.0          0       0   
23900    2020-03-18           Puerto Rico    2860840.0          0       0   
23901    2020-03-19           Puerto Rico    2860840.0          0       0   
23902    2020-03-20           Puerto Rico    2860840.0          0       0   
23903    2020-03-21           Puerto Rico    2860840.0          0       0   
24085    2020-03-17                 Congo    5518092.0          0       0   
24086    2020-03-18                 Congo    5518092.0          0       0   
24087    2020-03-19                 Congo    5518092.0          0       0   
24088    2020-03-20                 Congo    5518092.0          0       0   
24089    2020-03-21                 Congo    5518092.0          0       0   
29359    2020-03-19               Bahamas     393248.0          0       0   
29360    2020-03-20               Bahamas     393248.0          0       0   
29361    2020-03-21               Bahamas     393248.0          0       0   
29363    2020-03-18                Gambia    2416664.0          0       0   
29364    2020-03-19                Gambia    2416664.0          0       0   
29365    2020-03-20                Gambia    2416664.0          0       0   
29366    2020-03-21                Gambia    2416664.0          0       0   

       deaths  recovered  mortality_rate  incident_rate  
1384        0          0             NaN            0.0  
3975        0          0             NaN            0.0  
6370        0          0             NaN            0.0  
11864       0          0             NaN            0.0  
11865       0          0             NaN            0.0  
11866       0          0             NaN            0.0  
12033       0          0             NaN            0.0  
12034       0          0             NaN            0.0  
12035       0          0             NaN            0.0  
12036       0          0             NaN            0.0  
13179       0          0             NaN            0.0  
14148       0          0             NaN            0.0  
18464       0          0             NaN            0.0  
19511       0          0             NaN            0.0  
23342       0          0             NaN            0.0  
23899       0          0             NaN            0.0  
23900       0          0             NaN            0.0  
23901       0          0             NaN            0.0  
23902       0          0             NaN            0.0  
23903       0          0             NaN            0.0  
24085       0          0             NaN            0.0  
24086       0          0             NaN            0.0  
24087       0          0             NaN            0.0  
24088       0          0             NaN            0.0  
24089       0 

In [99]:
df = df.fillna(0)
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 31466 entries, 0 to 32603
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   reported_date   31466 non-null  datetime64[ns]
 1   country_region  31466 non-null  object        
 2   population      31466 non-null  float64       
 3   confirmed       31466 non-null  int64         
 4   active          31466 non-null  int64         
 5   deaths          31466 non-null  int64         
 6   recovered       31466 non-null  int64         
 7   mortality_rate  31466 non-null  float64       
 8   incident_rate   31466 non-null  float64       
dtypes: datetime64[ns](1), float64(3), int64(4), object(1)
memory usage: 3.4+ MB


- Giờ khi đã giải quyết xong các giá trị thiếu. Ta đã có thể tiến đến bước tiếp theo

## Đặt ra câu hỏi từ dữ liệu

## Phân tích dữ liệu để trả lời câu hỏi

## Tổng kết và nhìn lại quá trình làm đồ án